# Data Exploration

## Imports

In [1]:
#Turn off pretty printing
%pprint

#Import packages
import pandas as pd

Pretty printing has been turned OFF


## Children's Literature Reviews
Read in JSON file with review data and convert to DataFrame to prep for analysis

In [2]:
df = pd.read_json('/Users/ashleyfeiler/Desktop/Goodreads_Data/goodreads_reviews_children.json', lines=True)

In [3]:
df.head()

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d  23310161   
1  8842281e1d1347389f2ab93d60773d4d  17290220   
2  8842281e1d1347389f2ab93d60773d4d   6954929   
3  8842281e1d1347389f2ab93d60773d4d    460548   
4  8842281e1d1347389f2ab93d60773d4d  11474551   

                          review_id  rating  \
0  f4b4b050f4be00e9283c92a814af2670       4   
1  22d424a2b0057b18fb6ecf017af7be92       5   
2  50ed4431c451d5677d98dd25ca8ec106       5   
3  1e4de11dd4fa4b7ffa59b6c69a6b28e9       5   
4  2065145714bf747083a1c9ce81d5c4fe       5   

                                         review_text  \
0                        Fun sequel to the original.   
1  One of my favorite books to read to my 5 year ...   
2  One of the best and most imaginative childrens...   
3  My daughter is loving this. Published in the 6...   
4                  A friend sent me this. Hilarious!   

                       date_added                    date_updated  \
0  Tue Nov 17 11:37:35 -0800 2015  Tue Nov 17 11:38:05 -0800 2015   
1  Sat Nov 08 08:54:03 -0800 2014  Wed Jan 25 13:56:12 -0800 2017   
2  Thu Oct 23 13:46:20 -0700 2014  Thu Oct 23 13:47:00 -0700 2014   
3  Mon Dec 02 10:43:59 -0800 2013  Wed Mar 22 11:47:25 -0700 2017   
4  Wed May 11 22:38:11 -0700 2011  Sun Jan 29 15:56:41 -0800 2012   

                          read_at                      started_at  n_votes  \
0                                                                        7   
1  Tue Jan 24 00:00:00 -0800 2017                                        4   
2                                                                        6   
3                                                                        5   
4  Wed May 11 00:00:00 -0700 2011  Wed May 11 00:00:00 -0700 2011        5   

   n_comments  
0           0  
1           0  
2           1  
3           4  
4           0

In [4]:
df.shape

(734640, 11)

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 734640 entries, 0 to 734639
Data columns (total 11 columns):
 #   Column        Non-Null Count   Dtype 
---  ------        --------------   ----- 
 0   user_id       734640 non-null  object
 1   book_id       734640 non-null  int64 
 2   review_id     734640 non-null  object
 3   rating        734640 non-null  int64 
 4   review_text   734640 non-null  object
 5   date_added    734640 non-null  object
 6   date_updated  734640 non-null  object
 7   read_at       734640 non-null  object
 8   started_at    734640 non-null  object
 9   n_votes       734640 non-null  int64 
 10  n_comments    734640 non-null  int64 
dtypes: int64(4), object(7)
memory usage: 61.7+ MB


The number of reviews (734640) matches the number provided by the data source page, which is great! We will keep an eye on this number as we add to the DataFrame to make sure we aren't accidentally adding/removing anything. There are no NaN values, which is great, but there are lot of columns whose data is unecessary for this product, so let's trim that down to the essentials. 

In [6]:
reviews_df = df[['user_id', 'book_id', 'review_id', 'rating', 'review_text']]

In [7]:
reviews_df.head()

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d  23310161   
1  8842281e1d1347389f2ab93d60773d4d  17290220   
2  8842281e1d1347389f2ab93d60773d4d   6954929   
3  8842281e1d1347389f2ab93d60773d4d    460548   
4  8842281e1d1347389f2ab93d60773d4d  11474551   

                          review_id  rating  \
0  f4b4b050f4be00e9283c92a814af2670       4   
1  22d424a2b0057b18fb6ecf017af7be92       5   
2  50ed4431c451d5677d98dd25ca8ec106       5   
3  1e4de11dd4fa4b7ffa59b6c69a6b28e9       5   
4  2065145714bf747083a1c9ce81d5c4fe       5   

                                         review_text  
0                        Fun sequel to the original.  
1  One of my favorite books to read to my 5 year ...  
2  One of the best and most imaginative childrens...  
3  My daughter is loving this. Published in the 6...  
4                  A friend sent me this. Hilarious!

Much better! Now this data contains the book IDs, but we'd like to be able to identify the books by title/author and have some basic information such as page count. This book-specific data is in another JSON file, so let's load that into another DataFrame.

## Children's Literature Book Info

In [8]:
books = pd.read_json('/Users/ashleyfeiler/Desktop/Goodreads_Data/goodreads_books_children.json', lines=True)

In [9]:
books.head()

isbn  text_reviews_count    series country_code language_code  \
0  1599150603                   7        []           US                 
1  1934876569                   6  [151854]           US                 
2  0590417010                 193        []           US           eng   
3  0915190575                   4        []           US                 
4  1416904999                   4        []           US                 

                                     popular_shelves asin is_ebook  \
0  [{'count': '56', 'name': 'to-read'}, {'count':...         false   
1  [{'count': '515', 'name': 'to-read'}, {'count'...         false   
2  [{'count': '450', 'name': 'to-read'}, {'count'...         false   
3  [{'count': '8', 'name': 'to-read'}, {'count': ...         false   
4  [{'count': '8', 'name': 'to-read'}, {'count': ...         false   

   average_rating kindle_asin  ... publication_month edition_information  \
0            4.13  B00DU10PUG  ...                 9                       
1            4.22              ...                 3                       
2            4.43  B017RORXNI  ...                 9                       
3            4.29              ...                                         
4            3.57              ...                 6                       

  publication_year                                                url  \
0             2006  https://www.goodreads.com/book/show/287141.The...   
1             2009  https://www.goodreads.com/book/show/6066812-al...   
2             1995  https://www.goodreads.com/book/show/89378.Dog_...   
3                   https://www.goodreads.com/book/show/3209312-mo...   
4             2005  https://www.goodreads.com/book/show/1698376.Wh...   

                                           image_url  book_id ratings_count  \
0  https://s.gr-assets.com/assets/nophoto/book/11...   287141            46   
1  https://images.gr-assets.com/books/1316637798m...  6066812            98   
2  https://images.gr-assets.com/books/1360057676m...    89378          1331   
3  https://s.gr-assets.com/assets/nophoto/book/11...  3209312            11   
4  https://s.gr-assets.com/assets/nophoto/book/11...  1698376            23   

   work_id                                              title  \
0   278578                      The Aeneid for Boys and Girls   
1   701117  All's Fairy in Love and War (Avalon: Web of Ma...   
2    86259                                         Dog Heaven   
3  3242879  Moths and Mothers, Feathers and Fathers: A Sto...   
4  1695373                                    What Do You Do?   

                                title_without_series  
0                      The Aeneid for Boys and Girls  
1  All's Fairy in Love and War (Avalon: Web of Ma...  
2                                         Dog Heaven  
3  Moths and Mothers, Feathers and Fathers: A Sto...  
4                                    What Do You Do?  

[5 rows x 29 columns]

In [10]:
books.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 124082 entries, 0 to 124081
Data columns (total 29 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   isbn                  124082 non-null  object 
 1   text_reviews_count    124082 non-null  int64  
 2   series                124082 non-null  object 
 3   country_code          124082 non-null  object 
 4   language_code         124082 non-null  object 
 5   popular_shelves       124082 non-null  object 
 6   asin                  124082 non-null  object 
 7   is_ebook              124082 non-null  object 
 8   average_rating        124082 non-null  float64
 9   kindle_asin           124082 non-null  object 
 10  similar_books         124082 non-null  object 
 11  description           124082 non-null  object 
 12  format                124082 non-null  object 
 13  link                  124082 non-null  object 
 14  authors               124082 non-null  object 
 15  

This DataFrame once again has no NaN values (nice!), and while it does have a lower number of samples than the reviews DataFrame, this makes sense because there could be multiple reviews for the same book. However, this is a ton of info, so let's trim it down again to data we might need.

In [11]:
books_df = books[['text_reviews_count', 'series', 'country_code', 'language_code', 'is_ebook', 'average_rating', 'description', 'format', 'authors', 'publisher', 'num_pages', 'edition_information', 'publication_year', 'book_id', 'ratings_count', 'title', 'title_without_series']]

In [12]:
books_df.head()

text_reviews_count    series country_code language_code is_ebook  \
0                   7        []           US                  false   
1                   6  [151854]           US                  false   
2                 193        []           US           eng    false   
3                   4        []           US                  false   
4                   4        []           US                  false   

   average_rating                                        description  \
0            4.13  Relates in vigorous prose the tale of Aeneas, ...   
1            4.22  To Kara's astonishment, she discovers that a p...   
2            4.43  In Newbery Medalist Cynthia Rylant's classic b...   
3            4.29                                                      
4            3.57  WHAT DO YOU DO?\nA hen lays eggs...\nA cow giv...   

       format                                            authors  \
0   Paperback             [{'author_id': '3041852', 'role': ''}]   
1   Paperback               [{'author_id': '19158', 'role': ''}]   
2   Hardcover                [{'author_id': '5411', 'role': ''}]   
3              [{'author_id': '589328', 'role': ''}, {'author...   
4  Board Book              [{'author_id': '169159', 'role': ''}]   

              publisher num_pages edition_information publication_year  \
0  Yesterday's Classics       162                                 2006   
1            Seven Seas       216                                 2009   
2        Blue Sky Press        40                                 1995   
3                                                                        
4          Little Simon        24                                 2005   

   book_id  ratings_count                                              title  \
0   287141             46                      The Aeneid for Boys and Girls   
1  6066812             98  All's Fairy in Love and War (Avalon: Web of Ma...   
2    89378           1331                                         Dog Heaven   
3  3209312             11  Moths and Mothers, Feathers and Fathers: A Sto...   
4  1698376             23                                    What Do You Do?   

                                title_without_series  
0                      The Aeneid for Boys and Girls  
1  All's Fairy in Love and War (Avalon: Web of Ma...  
2                                         Dog Heaven  
3  Moths and Mothers, Feathers and Fathers: A Sto...  
4                                    What Do You Do?

In [13]:
books_df.tail()

text_reviews_count series country_code language_code is_ebook  \
124077                   1     []           US                   true   
124078                   7     []           US         en-GB     true   
124079                  13     []           US                  false   
124080                   1     []           US                  false   
124081                   6     []           US                  false   

        average_rating                                        description  \
124077            4.00                                                      
124078            4.27  Another hilarious and moving novel from David ...   
124079            4.29  "A perfect reminder to always be on the lookou...   
124080            4.36  One of the most popular series ever published ...   
124081            4.00  Gathers poems by William Blake, Emily Bronte, ...   

           format                                            authors  \
124077                         [{'author_id': '421368', 'role': ''}]   
124078                         [{'author_id': '313653', 'role': ''}]   
124079  Hardcover  [{'author_id': '1221698', 'role': ''}, {'autho...   
124080  Paperback  [{'author_id': '10681', 'role': ''}, {'author_...   
124081  Hardcover         [{'author_id': '82312', 'role': 'Editor'}]   

                           publisher num_pages edition_information  \
124077                                                               
124078  HarperCollinsChildren'sBooks       205                       
124079                 HarperCollins        40                       
124080                       Aladdin       176                       
124081          Smithmark Publishers        96                       

       publication_year   book_id  ratings_count  \
124077                   24558598              5   
124078             2011  13057172            839   
124079             2010   7925060             40   
124080             2002    331839             18   
124081             1996   2342551             36   

                                                 title  \
124077         Children's Books: Oliver and the Dragon   
124078                                  Gangsta Granny   
124079                                    Instructions   
124080  Jacqueline Kennedy Onassis: Friend of the Arts   
124081        The Children's Classic Poetry Collection   

                                  title_without_series  
124077         Children's Books: Oliver and the Dragon  
124078                                  Gangsta Granny  
124079                                    Instructions  
124080  Jacqueline Kennedy Onassis: Friend of the Arts  
124081        The Children's Classic Poetry Collection

Let's explore some of these columns.

In [14]:
books_df.country_code.value_counts()

US    124082
Name: country_code, dtype: int64

Looks like every book has the same country code - we can probably leave that column out!

In [15]:
books_df.language_code.value_counts()

         72496
eng      30253
en-US     6197
en-GB     2071
spa       1501
         ...  
egy          1
aze          1
mon          1
vls          1
nub          1
Name: language_code, Length: 90, dtype: int64

This will be important to consider - to scale down the data, I will likely restrict the data to only reviews of books written in English (more on this later!)

In [16]:
books_df.format.value_counts()

Hardcover                   51159
Paperback                   38880
                            21039
Board Book                   3214
ebook                        2423
                            ...  
pop up book                     1
miekka (paperback)              1
Klappenbroschur                 1
three books in boxed set        1
Board Book Periodical           1
Name: format, Length: 209, dtype: int64

209 different formats of books? I have no idea what those could be, and the book format likely doesn't influence the reviews that significantly, so I can probably leave out this column as well.

In [17]:
books_df.is_ebook.value_counts()

false    114606
true       9476
Name: is_ebook, dtype: int64

There are significantly more physical books represented in this data than ebooks.

In [18]:
books_df.loc[books_df['book_id'] == 23310161]

text_reviews_count    series country_code language_code is_ebook  \
17550                1231  [751058]           US           eng    false   

       average_rating                                        description  \
17550            4.43  The companion to the #1 blockbuster bestseller...   

          format                                            authors  \
17550  Hardcover  [{'author_id': '6561846', 'role': ''}, {'autho...   

            publisher num_pages edition_information publication_year  \
17550  Philomel Books        36                                 2015   

        book_id  ratings_count                          title  \
17550  23310161           8924  The Day the Crayons Came Home   

                title_without_series  
17550  The Day the Crayons Came Home

Indexing seems to work!
Now that I've explored the data a bit more, I'm going to remove a few more columns.

In [19]:
books_df = books[['text_reviews_count', 'series', 'language_code', 'is_ebook', 'average_rating', 'description', 'authors', 'publisher', 'num_pages', 'edition_information', 'publication_year', 'book_id', 'ratings_count', 'title', 'title_without_series']]

In [20]:
books_df.head()

text_reviews_count    series language_code is_ebook  average_rating  \
0                   7        []                  false            4.13   
1                   6  [151854]                  false            4.22   
2                 193        []           eng    false            4.43   
3                   4        []                  false            4.29   
4                   4        []                  false            3.57   

                                         description  \
0  Relates in vigorous prose the tale of Aeneas, ...   
1  To Kara's astonishment, she discovers that a p...   
2  In Newbery Medalist Cynthia Rylant's classic b...   
3                                                      
4  WHAT DO YOU DO?\nA hen lays eggs...\nA cow giv...   

                                             authors             publisher  \
0             [{'author_id': '3041852', 'role': ''}]  Yesterday's Classics   
1               [{'author_id': '19158', 'role': ''}]            Seven Seas   
2                [{'author_id': '5411', 'role': ''}]        Blue Sky Press   
3  [{'author_id': '589328', 'role': ''}, {'author...                         
4              [{'author_id': '169159', 'role': ''}]          Little Simon   

  num_pages edition_information publication_year  book_id  ratings_count  \
0       162                                 2006   287141             46   
1       216                                 2009  6066812             98   
2        40                                 1995    89378           1331   
3                                                 3209312             11   
4        24                                 2005  1698376             23   

                                               title  \
0                      The Aeneid for Boys and Girls   
1  All's Fairy in Love and War (Avalon: Web of Ma...   
2                                         Dog Heaven   
3  Moths and Mothers, Feathers and Fathers: A Sto...   
4                                    What Do You Do?   

                                title_without_series  
0                      The Aeneid for Boys and Girls  
1  All's Fairy in Love and War (Avalon: Web of Ma...  
2                                         Dog Heaven  
3  Moths and Mothers, Feathers and Fathers: A Sto...  
4                                    What Do You Do?

Now that I have the preliminary information I want, I'll use pandas' `.merge()` function to combine the review and book data, which will automatically align based off of their shared book_id values.

In [21]:
children_df = pd.merge(reviews_df, books_df)

In [22]:
children_df.head()

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d  23310161   
1  d1e368a7d2870eb6fbf6e0d350568a2d  23310161   
2  7370e3e727f4fdae3ab82f054838d73d  23310161   
3  ec58d8af7f69e19781abed9d25287a57  23310161   
4  cfe4738a5d926fc158d4fcb1287fe2c5  23310161   

                          review_id  rating  \
0  f4b4b050f4be00e9283c92a814af2670       4   
1  87fb7685b0b1aa774cc14c330bdcfadf       4   
2  b8c59999e366608b789114d6fdd52a69       5   
3  db7984daa9a8c3b49f4c7f9f73ec86ac       5   
4  f3c100afbe1495fc51bff4d6c3cb7138       4   

                                         review_text  text_reviews_count  \
0                        Fun sequel to the original.                1231   
1  Great sequel! Lost crayons write their owner p...                1231   
2                      Great sequel to the original.                1231   
3          I loved this book as much as the 1st one.                1231   
4  It was a cute follow-up to the clever first book.                1231   

     series language_code is_ebook  average_rating  \
0  [751058]           eng    false            4.43   
1  [751058]           eng    false            4.43   
2  [751058]           eng    false            4.43   
3  [751058]           eng    false            4.43   
4  [751058]           eng    false            4.43   

                                         description  \
0  The companion to the #1 blockbuster bestseller...   
1  The companion to the #1 blockbuster bestseller...   
2  The companion to the #1 blockbuster bestseller...   
3  The companion to the #1 blockbuster bestseller...   
4  The companion to the #1 blockbuster bestseller...   

                                             authors       publisher  \
0  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   
1  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   
2  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   
3  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   
4  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   

  num_pages edition_information publication_year  ratings_count  \
0        36                                 2015           8924   
1        36                                 2015           8924   
2        36                                 2015           8924   
3        36                                 2015           8924   
4        36                                 2015           8924   

                           title           title_without_series  
0  The Day the Crayons Came Home  The Day the Crayons Came Home  
1  The Day the Crayons Came Home  The Day the Crayons Came Home  
2  The Day the Crayons Came Home  The Day the Crayons Came Home  
3  The Day the Crayons Came Home  The Day the Crayons Came Home  
4  The Day the Crayons Came Home  The Day the Crayons Came Home

In [23]:
children_df.shape

(734640, 19)

We still have 734640 rows, meaning all the samples are accounted for, and the columns have increased given that we've combined the information from both DataFrames. Now I'm going to take a closer look at some of the text just to make sure it's there in full (I'll dive into analysis of how the text is represented later on). 

In [24]:
children_df.review_text[1]

"Great sequel! Lost crayons write their owner postcards from where they've been lost so they can get retrieved. Some of them have been through so much! \n Good for preschool storytimes."

In [25]:
children_df.review_text[10000]

'I\'m torn. Do I read it again immediately, or wait a few months? I need to do so at some point, because reading for the adventure & character dev. wasn\'t nearly enough. \n I love Oscar, and Callie, and the cats. I love the way Oscar (who seems slightly autistic) has an understanding of himself and the world that unfolds at the same pace for him as it did for me. I\'m truly spellbound by the poetic language and mysterious atmosphere. (Opening the book at a random early page finds Oscar in his master\'s library: "The shelves were so very much taller than he could ever dream of being, and Oscar firmly believed people shouldn\'t go any higher than they already were.") \n Like Breadcrumbs it\'s not a book for everyone. But it\'s brilliant. Imo, Ursu (so far) writes with almost as much grace & power as Sonya Hartnett and her stories are a lot easier to stomach. I will definitely read more by the author.'

## Children's Literature Genre Info
This is the children's genre subset of data, but the data provided offers multiple possible genre tags for each book, which could be useful looking at potential overlap between categories. Once again, genre info is in a different JSON file, so I'll load that here and follow the same process as with the other DataFrames so far, merging based on shared book IDs.

In [26]:
genres_df = pd.read_json('/Users/ashleyfeiler/Desktop/Goodreads_Data/goodreads_book_genres_initial.json', lines=True)

In [27]:
genres_df.head()

book_id                                             genres
0  5333265      {'history, historical fiction, biography': 1}
1  1333909  {'fiction': 219, 'history, historical fiction,...
2  7327624  {'fantasy, paranormal': 31, 'fiction': 8, 'mys...
3  6066819  {'fiction': 555, 'romance': 23, 'mystery, thri...
4   287140                                 {'non-fiction': 3}

In [28]:
genres_df.loc[genres_df['book_id'] == 287140]

book_id              genres
4   287140  {'non-fiction': 3}

In [29]:
children_df = pd.merge(children_df, genres_df)
children_df['Genre_Tag'] = 'children'

Here I've merged the list of genres associated with each book, but have also broadcast the genre tag 'children' to every entry in this category to represent that it is part of the children's literature genre subset. This will be helpful for making a DataFrame with all the different genres combined to keep different genres differentiated.

In [30]:
children_df.head()

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d  23310161   
1  d1e368a7d2870eb6fbf6e0d350568a2d  23310161   
2  7370e3e727f4fdae3ab82f054838d73d  23310161   
3  ec58d8af7f69e19781abed9d25287a57  23310161   
4  cfe4738a5d926fc158d4fcb1287fe2c5  23310161   

                          review_id  rating  \
0  f4b4b050f4be00e9283c92a814af2670       4   
1  87fb7685b0b1aa774cc14c330bdcfadf       4   
2  b8c59999e366608b789114d6fdd52a69       5   
3  db7984daa9a8c3b49f4c7f9f73ec86ac       5   
4  f3c100afbe1495fc51bff4d6c3cb7138       4   

                                         review_text  text_reviews_count  \
0                        Fun sequel to the original.                1231   
1  Great sequel! Lost crayons write their owner p...                1231   
2                      Great sequel to the original.                1231   
3          I loved this book as much as the 1st one.                1231   
4  It was a cute follow-up to the clever first book.                1231   

     series language_code is_ebook  average_rating  ...  \
0  [751058]           eng    false            4.43  ...   
1  [751058]           eng    false            4.43  ...   
2  [751058]           eng    false            4.43  ...   
3  [751058]           eng    false            4.43  ...   
4  [751058]           eng    false            4.43  ...   

                                             authors       publisher  \
0  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   
1  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   
2  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   
3  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   
4  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   

  num_pages edition_information publication_year ratings_count  \
0        36                                 2015          8924   
1        36                                 2015          8924   
2        36                                 2015          8924   
3        36                                 2015          8924   
4        36                                 2015          8924   

                           title           title_without_series  \
0  The Day the Crayons Came Home  The Day the Crayons Came Home   
1  The Day the Crayons Came Home  The Day the Crayons Came Home   
2  The Day the Crayons Came Home  The Day the Crayons Came Home   
3  The Day the Crayons Came Home  The Day the Crayons Came Home   
4  The Day the Crayons Came Home  The Day the Crayons Came Home   

                                              genres Genre_Tag  
0  {'children': 931, 'fiction': 64, 'fantasy, par...  children  
1  {'children': 931, 'fiction': 64, 'fantasy, par...  children  
2  {'children': 931, 'fiction': 64, 'fantasy, par...  children  
3  {'children': 931, 'fiction': 64, 'fantasy, par...  children  
4  {'children': 931, 'fiction': 64, 'fantasy, par...  children  

[5 rows x 21 columns]

## Children's Literature Author Info
So far the children_df DataFrame has everything except author names, which is represented by an author_id deeply embedded in the 'authors' column. The format of the 'authors' column is formatted strangely, so let's take a closer look at the first row's entry.

In [31]:
children_df.authors[0]

[{'author_id': '6561846', 'role': ''}, {'author_id': '173545', 'role': 'Illustrations'}]

The information in the 'authors' column is a list of two dictionaries, each with two key:value pairs for the author and illustrator. I only want the first key:value pair of the first dictionary for the author's name, but it will probably take some digging to get to it. For now, I'm going to load the JSON file with author-specific info.

In [32]:
authors_df = pd.read_json('/Users/ashleyfeiler/Desktop/Goodreads_Data/goodreads_book_authors.json', lines=True)

In [33]:
authors_df.head()

average_rating  author_id  text_reviews_count              name  \
0            3.98     604031                   7  Ronald J. Fields   
1            4.08     626222               28716     Anita Diamant   
2            3.92      10333                5075    Barbara Hambly   
3            3.68       9212               36262   Jennifer Weiner   
4            3.82     149918                  96     Nigel Pennick   

   ratings_count  
0             49  
1         546796  
2         122118  
3         888522  
4           1740

Now that I've got the author DataFrame, I know I will need to merge based on the author_id, so let's try to extract the author_id from the main DataFrame. I'm not sure how this is going to work, so I'm going to try it on a small sample first.

In [34]:
test_authors = children_df.head()

In [35]:
test_authors

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d  23310161   
1  d1e368a7d2870eb6fbf6e0d350568a2d  23310161   
2  7370e3e727f4fdae3ab82f054838d73d  23310161   
3  ec58d8af7f69e19781abed9d25287a57  23310161   
4  cfe4738a5d926fc158d4fcb1287fe2c5  23310161   

                          review_id  rating  \
0  f4b4b050f4be00e9283c92a814af2670       4   
1  87fb7685b0b1aa774cc14c330bdcfadf       4   
2  b8c59999e366608b789114d6fdd52a69       5   
3  db7984daa9a8c3b49f4c7f9f73ec86ac       5   
4  f3c100afbe1495fc51bff4d6c3cb7138       4   

                                         review_text  text_reviews_count  \
0                        Fun sequel to the original.                1231   
1  Great sequel! Lost crayons write their owner p...                1231   
2                      Great sequel to the original.                1231   
3          I loved this book as much as the 1st one.                1231   
4  It was a cute follow-up to the clever first book.                1231   

     series language_code is_ebook  average_rating  ...  \
0  [751058]           eng    false            4.43  ...   
1  [751058]           eng    false            4.43  ...   
2  [751058]           eng    false            4.43  ...   
3  [751058]           eng    false            4.43  ...   
4  [751058]           eng    false            4.43  ...   

                                             authors       publisher  \
0  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   
1  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   
2  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   
3  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   
4  [{'author_id': '6561846', 'role': ''}, {'autho...  Philomel Books   

  num_pages edition_information publication_year ratings_count  \
0        36                                 2015          8924   
1        36                                 2015          8924   
2        36                                 2015          8924   
3        36                                 2015          8924   
4        36                                 2015          8924   

                           title           title_without_series  \
0  The Day the Crayons Came Home  The Day the Crayons Came Home   
1  The Day the Crayons Came Home  The Day the Crayons Came Home   
2  The Day the Crayons Came Home  The Day the Crayons Came Home   
3  The Day the Crayons Came Home  The Day the Crayons Came Home   
4  The Day the Crayons Came Home  The Day the Crayons Came Home   

                                              genres Genre_Tag  
0  {'children': 931, 'fiction': 64, 'fantasy, par...  children  
1  {'children': 931, 'fiction': 64, 'fantasy, par...  children  
2  {'children': 931, 'fiction': 64, 'fantasy, par...  children  
3  {'children': 931, 'fiction': 64, 'fantasy, par...  children  
4  {'children': 931, 'fiction': 64, 'fantasy, par...  children  

[5 rows x 21 columns]

In [36]:
test_authors.authors.iloc[2][0]

{'author_id': '6561846', 'role': ''}

In [37]:
test_list = list(test_authors.index.map(lambda x: test_authors.authors.iloc[x][0]))
list(test_list[0].values())[0]

'6561846'

In [38]:
test_authors['author_id'] = test_authors.index.map(lambda x: list(test_list[x].values())[0])

/var/folders/j9/9lmff3zj6bxgfpsb8gm8pjrw0000gn/T/ipykernel_19532/2851788888.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  test_authors['author_id'] = test_authors.index.map(lambda x: list(test_list[x].values())[0])


In [39]:
test_authors

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d  23310161   
1  d1e368a7d2870eb6fbf6e0d350568a2d  23310161   
2  7370e3e727f4fdae3ab82f054838d73d  23310161   
3  ec58d8af7f69e19781abed9d25287a57  23310161   
4  cfe4738a5d926fc158d4fcb1287fe2c5  23310161   

                          review_id  rating  \
0  f4b4b050f4be00e9283c92a814af2670       4   
1  87fb7685b0b1aa774cc14c330bdcfadf       4   
2  b8c59999e366608b789114d6fdd52a69       5   
3  db7984daa9a8c3b49f4c7f9f73ec86ac       5   
4  f3c100afbe1495fc51bff4d6c3cb7138       4   

                                         review_text  text_reviews_count  \
0                        Fun sequel to the original.                1231   
1  Great sequel! Lost crayons write their owner p...                1231   
2                      Great sequel to the original.                1231   
3          I loved this book as much as the 1st one.                1231   
4  It was a cute follow-up to the clever first book.                1231   

     series language_code is_ebook  average_rating  ...       publisher  \
0  [751058]           eng    false            4.43  ...  Philomel Books   
1  [751058]           eng    false            4.43  ...  Philomel Books   
2  [751058]           eng    false            4.43  ...  Philomel Books   
3  [751058]           eng    false            4.43  ...  Philomel Books   
4  [751058]           eng    false            4.43  ...  Philomel Books   

  num_pages edition_information publication_year ratings_count  \
0        36                                 2015          8924   
1        36                                 2015          8924   
2        36                                 2015          8924   
3        36                                 2015          8924   
4        36                                 2015          8924   

                           title           title_without_series  \
0  The Day the Crayons Came Home  The Day the Crayons Came Home   
1  The Day the Crayons Came Home  The Day the Crayons Came Home   
2  The Day the Crayons Came Home  The Day the Crayons Came Home   
3  The Day the Crayons Came Home  The Day the Crayons Came Home   
4  The Day the Crayons Came Home  The Day the Crayons Came Home   

                                              genres Genre_Tag author_id  
0  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846  
1  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846  
2  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846  
3  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846  
4  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846  

[5 rows x 22 columns]

This seems to work, time to do it to all the data!

In [40]:
author_list = list(children_df.index.map(lambda x: children_df.authors.iloc[x][0]))
list(author_list[0].values())[0]

'6561846'

In [41]:
children_df['author_id'] = children_df.index.map(lambda x: int(list(author_list[x].values())[0]))
children_df.head()

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d  23310161   
1  d1e368a7d2870eb6fbf6e0d350568a2d  23310161   
2  7370e3e727f4fdae3ab82f054838d73d  23310161   
3  ec58d8af7f69e19781abed9d25287a57  23310161   
4  cfe4738a5d926fc158d4fcb1287fe2c5  23310161   

                          review_id  rating  \
0  f4b4b050f4be00e9283c92a814af2670       4   
1  87fb7685b0b1aa774cc14c330bdcfadf       4   
2  b8c59999e366608b789114d6fdd52a69       5   
3  db7984daa9a8c3b49f4c7f9f73ec86ac       5   
4  f3c100afbe1495fc51bff4d6c3cb7138       4   

                                         review_text  text_reviews_count  \
0                        Fun sequel to the original.                1231   
1  Great sequel! Lost crayons write their owner p...                1231   
2                      Great sequel to the original.                1231   
3          I loved this book as much as the 1st one.                1231   
4  It was a cute follow-up to the clever first book.                1231   

     series language_code is_ebook  average_rating  ...       publisher  \
0  [751058]           eng    false            4.43  ...  Philomel Books   
1  [751058]           eng    false            4.43  ...  Philomel Books   
2  [751058]           eng    false            4.43  ...  Philomel Books   
3  [751058]           eng    false            4.43  ...  Philomel Books   
4  [751058]           eng    false            4.43  ...  Philomel Books   

  num_pages edition_information publication_year ratings_count  \
0        36                                 2015          8924   
1        36                                 2015          8924   
2        36                                 2015          8924   
3        36                                 2015          8924   
4        36                                 2015          8924   

                           title           title_without_series  \
0  The Day the Crayons Came Home  The Day the Crayons Came Home   
1  The Day the Crayons Came Home  The Day the Crayons Came Home   
2  The Day the Crayons Came Home  The Day the Crayons Came Home   
3  The Day the Crayons Came Home  The Day the Crayons Came Home   
4  The Day the Crayons Came Home  The Day the Crayons Came Home   

                                              genres Genre_Tag author_id  
0  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846  
1  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846  
2  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846  
3  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846  
4  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846  

[5 rows x 22 columns]

In [42]:
children_df.shape

(734640, 22)

In [43]:
authors_df = authors_df[['author_id', 'name']]
authors_df.head()

author_id              name
0     604031  Ronald J. Fields
1     626222     Anita Diamant
2      10333    Barbara Hambly
3       9212   Jennifer Weiner
4     149918     Nigel Pennick

In [44]:
children_df = pd.merge(children_df, authors_df)
children_df.shape

(734640, 23)

In [45]:
children_df.head()

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d  23310161   
1  d1e368a7d2870eb6fbf6e0d350568a2d  23310161   
2  7370e3e727f4fdae3ab82f054838d73d  23310161   
3  ec58d8af7f69e19781abed9d25287a57  23310161   
4  cfe4738a5d926fc158d4fcb1287fe2c5  23310161   

                          review_id  rating  \
0  f4b4b050f4be00e9283c92a814af2670       4   
1  87fb7685b0b1aa774cc14c330bdcfadf       4   
2  b8c59999e366608b789114d6fdd52a69       5   
3  db7984daa9a8c3b49f4c7f9f73ec86ac       5   
4  f3c100afbe1495fc51bff4d6c3cb7138       4   

                                         review_text  text_reviews_count  \
0                        Fun sequel to the original.                1231   
1  Great sequel! Lost crayons write their owner p...                1231   
2                      Great sequel to the original.                1231   
3          I loved this book as much as the 1st one.                1231   
4  It was a cute follow-up to the clever first book.                1231   

     series language_code is_ebook  average_rating  ... num_pages  \
0  [751058]           eng    false            4.43  ...        36   
1  [751058]           eng    false            4.43  ...        36   
2  [751058]           eng    false            4.43  ...        36   
3  [751058]           eng    false            4.43  ...        36   
4  [751058]           eng    false            4.43  ...        36   

  edition_information publication_year ratings_count  \
0                                 2015          8924   
1                                 2015          8924   
2                                 2015          8924   
3                                 2015          8924   
4                                 2015          8924   

                           title           title_without_series  \
0  The Day the Crayons Came Home  The Day the Crayons Came Home   
1  The Day the Crayons Came Home  The Day the Crayons Came Home   
2  The Day the Crayons Came Home  The Day the Crayons Came Home   
3  The Day the Crayons Came Home  The Day the Crayons Came Home   
4  The Day the Crayons Came Home  The Day the Crayons Came Home   

                                              genres Genre_Tag author_id  \
0  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   
1  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   
2  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   
3  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   
4  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   

           name  
0  Drew Daywalt  
1  Drew Daywalt  
2  Drew Daywalt  
3  Drew Daywalt  
4  Drew Daywalt  

[5 rows x 23 columns]

## Language Filtering
Given the information about all the different language codes before, I want to filter out only those books written in Enlglish. First lets look at a list of all the different language codes.

In [46]:
langs = list(children_df.language_code.unique())
langs

['eng', '', 'vie', 'spa', 'por', 'dan', 'en-US', 'ger', 'afr', 'mul', 'fin', 'en-GB', 'per', 'ara', 'rum', 'cat', 'nl', 'tur', 'pes', 'ita', 'ind', 'fre', 'srp', 'gre', 'nor', 'slo', 'swe', 'msa', 'cze', 'bul', 'lit', 'rus', 'ben', 'pol', 'mon', 'scr', 'heb', 'kat', 'kur', 'sin', 'en-CA', 'hye', 'lav', 'slv', 'pt-BR', 'est', 'isl', 'tha', 'hun', 'nob', 'ukr', 'jpn', 'gle', 'epo', 'oci', 'aze', 'zho', '--', 'kaz', 'mkd', 'krl', 'kor', 'lat', 'mal', 'sco', 'dum', 'eus', 'yid', 'roh', 'fao', 'hmn', 'egy', 'hin', 'fil', 'smn', 'dut', 'bos', 'tgl', 'mlt', 'nno', 'glg', 'guj', 'sqi', 'nub', 'vls', 'nld', 'en', 'gem', 'nav', 'non']

Most of these are 3-letter standardized language codes, but some are unrecognized. I'm going to take a look at the titles of books with unrecognized codes to make sure I'm not missing anything.

In [47]:
children_df.loc[children_df['language_code'] == 'scr'].title.iloc[:10] #Unknown?

9363                         Sofijin svijet
9364                         Sofijin svijet
17194                                  Čudo
17195                                  Čudo
17196                                  Čudo
31741    Do grla u nevolji (39 tragova, #6)
31742         Kao u grobu (39 tragova,  #4)
80766                            Mali princ
80915                            Mali princ
80916                            Mali princ
Name: title, dtype: object

In [48]:
children_df.loc[children_df['language_code'] == 'sco'].title.iloc[:10] #Scots

197760                                   The Doric Gruffalo
197971                                The Shetland Gruffalo
418080    Precious and the Puggies: Precious Ramotswe's ...
418081    Precious and the Puggies: Precious Ramotswe's ...
418082    Precious and the Puggies: Precious Ramotswe's ...
418083    Precious and the Puggies: Precious Ramotswe's ...
418084    Precious and the Puggies: Precious Ramotswe's ...
418174    Precious and the Puggies: Precious Ramotswe's ...
515260                                        Snotty Crocky
515261                                        Snotty Crocky
Name: title, dtype: object

In [49]:
children_df.loc[children_df['language_code'] == 'pes'].title.iloc[:10] #Iranian Persian

7184      سفر به سرزمین وحشی‌ها
17130                    اعجوبه
42212         راز موتورسیکلت من
42213         راز موتورسیکلت من
42214         راز موتورسیکلت من
116857              ماجرای سیرک
171689                سارا کورو
171690                سارا کورو
176707      دوازده پرنسس رقصنده
213234       استنلی، پسرک کاغذی
Name: title, dtype: object

In [50]:
children_df.loc[children_df['language_code'] == 'non'].title.iloc[:10] #Unknown?

726902    Vivaldi
Name: title, dtype: object

In [51]:
children_df.loc[children_df['language_code'] == 'vls'].title.iloc[:10] #West Flemish

707300    Mein 24. Dezember. Eine seltsame Geschichte
Name: title, dtype: object

In [52]:
children_df.loc[children_df['language_code'] == ''].title.iloc[-10:] #Unknown?

734619                                      The Witch-Child
734620                    Peter Rabbit Little Learning Book
734621    The Christian Mom's Idea Book: Hundreds of Ide...
734624    The Guardian Princesses Inaugural Trilogy: Spe...
734625                               Nobody Has Time for Me
734630                           When Teddy Was In The Wash
734631                                  The Lost Tooth Club
734632      A Journey The Autobiography of Apolo Anton Ohno
734634                                     Work, Work, Work
734635                                    Where Peace Lives
Name: title, dtype: object

In [53]:
children_df.loc[children_df['language_code'] == '--'].title.iloc[:10] #Unknown?

81175                                    El Little Principe
482539    All of a Kind Family: Five Young Sisters in th...
482540    All of a Kind Family: Five Young Sisters in th...
724333    Mein Körper gehört mir! Schutz vor Missbrauch ...
Name: title, dtype: object

While some of these language codes seem to include English titles, especially the empty string instead of a language code, to keep things consistent, I am going to filter only those books with some form of English as an assigned language code.

In [54]:
english = (children_df.language_code == 'eng') | (children_df.language_code == 'en-GB') | (children_df.language_code == 'en-US') | (children_df.language_code == 'en-CA')
children_df[english].head()

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d  23310161   
1  d1e368a7d2870eb6fbf6e0d350568a2d  23310161   
2  7370e3e727f4fdae3ab82f054838d73d  23310161   
3  ec58d8af7f69e19781abed9d25287a57  23310161   
4  cfe4738a5d926fc158d4fcb1287fe2c5  23310161   

                          review_id  rating  \
0  f4b4b050f4be00e9283c92a814af2670       4   
1  87fb7685b0b1aa774cc14c330bdcfadf       4   
2  b8c59999e366608b789114d6fdd52a69       5   
3  db7984daa9a8c3b49f4c7f9f73ec86ac       5   
4  f3c100afbe1495fc51bff4d6c3cb7138       4   

                                         review_text  text_reviews_count  \
0                        Fun sequel to the original.                1231   
1  Great sequel! Lost crayons write their owner p...                1231   
2                      Great sequel to the original.                1231   
3          I loved this book as much as the 1st one.                1231   
4  It was a cute follow-up to the clever first book.                1231   

     series language_code is_ebook  average_rating  ... num_pages  \
0  [751058]           eng    false            4.43  ...        36   
1  [751058]           eng    false            4.43  ...        36   
2  [751058]           eng    false            4.43  ...        36   
3  [751058]           eng    false            4.43  ...        36   
4  [751058]           eng    false            4.43  ...        36   

  edition_information publication_year ratings_count  \
0                                 2015          8924   
1                                 2015          8924   
2                                 2015          8924   
3                                 2015          8924   
4                                 2015          8924   

                           title           title_without_series  \
0  The Day the Crayons Came Home  The Day the Crayons Came Home   
1  The Day the Crayons Came Home  The Day the Crayons Came Home   
2  The Day the Crayons Came Home  The Day the Crayons Came Home   
3  The Day the Crayons Came Home  The Day the Crayons Came Home   
4  The Day the Crayons Came Home  The Day the Crayons Came Home   

                                              genres Genre_Tag author_id  \
0  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   
1  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   
2  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   
3  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   
4  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   

           name  
0  Drew Daywalt  
1  Drew Daywalt  
2  Drew Daywalt  
3  Drew Daywalt  
4  Drew Daywalt  

[5 rows x 23 columns]

In [55]:
children_df = children_df[english]
children_df.head()

user_id   book_id  \
0  8842281e1d1347389f2ab93d60773d4d  23310161   
1  d1e368a7d2870eb6fbf6e0d350568a2d  23310161   
2  7370e3e727f4fdae3ab82f054838d73d  23310161   
3  ec58d8af7f69e19781abed9d25287a57  23310161   
4  cfe4738a5d926fc158d4fcb1287fe2c5  23310161   

                          review_id  rating  \
0  f4b4b050f4be00e9283c92a814af2670       4   
1  87fb7685b0b1aa774cc14c330bdcfadf       4   
2  b8c59999e366608b789114d6fdd52a69       5   
3  db7984daa9a8c3b49f4c7f9f73ec86ac       5   
4  f3c100afbe1495fc51bff4d6c3cb7138       4   

                                         review_text  text_reviews_count  \
0                        Fun sequel to the original.                1231   
1  Great sequel! Lost crayons write their owner p...                1231   
2                      Great sequel to the original.                1231   
3          I loved this book as much as the 1st one.                1231   
4  It was a cute follow-up to the clever first book.                1231   

     series language_code is_ebook  average_rating  ... num_pages  \
0  [751058]           eng    false            4.43  ...        36   
1  [751058]           eng    false            4.43  ...        36   
2  [751058]           eng    false            4.43  ...        36   
3  [751058]           eng    false            4.43  ...        36   
4  [751058]           eng    false            4.43  ...        36   

  edition_information publication_year ratings_count  \
0                                 2015          8924   
1                                 2015          8924   
2                                 2015          8924   
3                                 2015          8924   
4                                 2015          8924   

                           title           title_without_series  \
0  The Day the Crayons Came Home  The Day the Crayons Came Home   
1  The Day the Crayons Came Home  The Day the Crayons Came Home   
2  The Day the Crayons Came Home  The Day the Crayons Came Home   
3  The Day the Crayons Came Home  The Day the Crayons Came Home   
4  The Day the Crayons Came Home  The Day the Crayons Came Home   

                                              genres Genre_Tag author_id  \
0  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   
1  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   
2  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   
3  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   
4  {'children': 931, 'fiction': 64, 'fantasy, par...  children   6561846   

           name  
0  Drew Daywalt  
1  Drew Daywalt  
2  Drew Daywalt  
3  Drew Daywalt  
4  Drew Daywalt  

[5 rows x 23 columns]

In [56]:
children_df.language_code.value_counts()

eng      326034
en-US     87456
en-GB     12892
en-CA      1244
Name: language_code, dtype: int64

In [57]:
children_df.shape

(427626, 23)

The shape and value_counts confirm that there are now only English-coded books and that the total number of reviews has now been reduced to 427626 - that's still a huge number of reviews, and we'll be working with mroe genres, so we're eventually going to have to take a sample of this data.

## Formatting
There are still a lot of columns in my working DataFrame, so I want to break it down into only the most essential pieces of data.

In [58]:
children_df.columns

Index(['user_id', 'book_id', 'review_id', 'rating', 'review_text',
       'text_reviews_count', 'series', 'language_code', 'is_ebook',
       'average_rating', 'description', 'authors', 'publisher', 'num_pages',
       'edition_information', 'publication_year', 'ratings_count', 'title',
       'title_without_series', 'genres', 'Genre_Tag', 'author_id', 'name'],
      dtype='object')

In [59]:
children_short = children_df[['user_id', 'rating', 'review_text', 'language_code', 'average_rating', 'num_pages', 'publication_year', 'ratings_count', 'title', 'genres', 'Genre_Tag']]
children_short.columns = ['User_ID', 'Rating', 'Text', 'Language', 'Avg_Rating', 'Pages', 'Pub_Year', 'Ratings_Count', 'Title', 'Genres', 'Author']

In [60]:
children_short.head()

User_ID  Rating  \
0  8842281e1d1347389f2ab93d60773d4d       4   
1  d1e368a7d2870eb6fbf6e0d350568a2d       4   
2  7370e3e727f4fdae3ab82f054838d73d       5   
3  ec58d8af7f69e19781abed9d25287a57       5   
4  cfe4738a5d926fc158d4fcb1287fe2c5       4   

                                                Text Language  Avg_Rating  \
0                        Fun sequel to the original.      eng        4.43   
1  Great sequel! Lost crayons write their owner p...      eng        4.43   
2                      Great sequel to the original.      eng        4.43   
3          I loved this book as much as the 1st one.      eng        4.43   
4  It was a cute follow-up to the clever first book.      eng        4.43   

  Pages Pub_Year  Ratings_Count                          Title  \
0    36     2015           8924  The Day the Crayons Came Home   
1    36     2015           8924  The Day the Crayons Came Home   
2    36     2015           8924  The Day the Crayons Came Home   
3    36     2015           8924  The Day the Crayons Came Home   
4    36     2015           8924  The Day the Crayons Came Home   

                                              Genres    Author  
0  {'children': 931, 'fiction': 64, 'fantasy, par...  children  
1  {'children': 931, 'fiction': 64, 'fantasy, par...  children  
2  {'children': 931, 'fiction': 64, 'fantasy, par...  children  
3  {'children': 931, 'fiction': 64, 'fantasy, par...  children  
4  {'children': 931, 'fiction': 64, 'fantasy, par...  children

In [61]:
children_short.tail()

User_ID  Rating  \
734633  897b4731bce9b5b20b8c75714b191c01       5   
734636  12104bd507c4ed850f90b22243905be0       3   
734637  f099c0c488290d550eff5200c64a3094       5   
734638  f099c0c488290d550eff5200c64a3094       4   
734639  0c6a008c66d219aa559aa717d48515a9       4   

                                                     Text Language  \
734633  My little ones loved this story, especially th...    en-GB   
734636  The Red Gorilla of Oz was alright. I was doing...      eng   
734637  A slightly more adult theme to this second in ...    en-GB   
734638  A delightful children's story that will appeal...    en-GB   
734639  This is a great book for creating a home or fo...      eng   

        Avg_Rating Pages Pub_Year  Ratings_Count  \
734633        3.78                             9   
734636        3.41   168     2012              6   
734637        5.00    73     2013              1   
734638        4.50    64     2013              2   
734639        3.94   114     2006             79   

                                           Title  \
734633  Mac The Fire Truck and The Airplane Fire   
734636                     The Red Gorilla of Oz   
734637             Tuscan Cats Get Into Mischief   
734638                 A Tale of Two Tuscan Cats   
734639   Teaching Yoga to Children Through Story   

                                           Genres    Author  
734633                            {'children': 2}  children  
734636  {'fantasy, paranormal': 2, 'children': 2}  children  
734637                            {'children': 2}  children  
734638                            {'children': 2}  children  
734639          {'non-fiction': 5, 'children': 3}  children

Let's rearrange these columns!

In [62]:
children_short = children_short[[ 'Text', 'Rating', 'Title', 'Author', 'Genres', 'Language', 'Pages', 'Pub_Year', 'Avg_Rating', 'Ratings_Count', 'User_ID']]
children_short.head()

Text  Rating  \
0                        Fun sequel to the original.       4   
1  Great sequel! Lost crayons write their owner p...       4   
2                      Great sequel to the original.       5   
3          I loved this book as much as the 1st one.       5   
4  It was a cute follow-up to the clever first book.       4   

                           Title    Author  \
0  The Day the Crayons Came Home  children   
1  The Day the Crayons Came Home  children   
2  The Day the Crayons Came Home  children   
3  The Day the Crayons Came Home  children   
4  The Day the Crayons Came Home  children   

                                              Genres Language Pages Pub_Year  \
0  {'children': 931, 'fiction': 64, 'fantasy, par...      eng    36     2015   
1  {'children': 931, 'fiction': 64, 'fantasy, par...      eng    36     2015   
2  {'children': 931, 'fiction': 64, 'fantasy, par...      eng    36     2015   
3  {'children': 931, 'fiction': 64, 'fantasy, par...      eng    36     2015   
4  {'children': 931, 'fiction': 64, 'fantasy, par...      eng    36     2015   

   Avg_Rating  Ratings_Count                           User_ID  
0        4.43           8924  8842281e1d1347389f2ab93d60773d4d  
1        4.43           8924  d1e368a7d2870eb6fbf6e0d350568a2d  
2        4.43           8924  7370e3e727f4fdae3ab82f054838d73d  
3        4.43           8924  ec58d8af7f69e19781abed9d25287a57  
4        4.43           8924  cfe4738a5d926fc158d4fcb1287fe2c5

In [63]:
children_short.tail()

Text  Rating  \
734633  My little ones loved this story, especially th...       5   
734636  The Red Gorilla of Oz was alright. I was doing...       3   
734637  A slightly more adult theme to this second in ...       5   
734638  A delightful children's story that will appeal...       4   
734639  This is a great book for creating a home or fo...       4   

                                           Title    Author  \
734633  Mac The Fire Truck and The Airplane Fire  children   
734636                     The Red Gorilla of Oz  children   
734637             Tuscan Cats Get Into Mischief  children   
734638                 A Tale of Two Tuscan Cats  children   
734639   Teaching Yoga to Children Through Story  children   

                                           Genres Language Pages Pub_Year  \
734633                            {'children': 2}    en-GB                  
734636  {'fantasy, paranormal': 2, 'children': 2}      eng   168     2012   
734637                            {'children': 2}    en-GB    73     2013   
734638                            {'children': 2}    en-GB    64     2013   
734639          {'non-fiction': 5, 'children': 3}      eng   114     2006   

        Avg_Rating  Ratings_Count                           User_ID  
734633        3.78              9  897b4731bce9b5b20b8c75714b191c01  
734636        3.41              6  12104bd507c4ed850f90b22243905be0  
734637        5.00              1  f099c0c488290d550eff5200c64a3094  
734638        4.50              2  f099c0c488290d550eff5200c64a3094  
734639        3.94             79  0c6a008c66d219aa559aa717d48515a9

In [64]:
children_short.shape

(427626, 11)

As mentioned before, to make this data more manageable, I'm going to to take a sample of 1000 reviews from each genre and save this shortened DataFrame with the information for those 1000 reviews into CSV files that I can then access later.

In [65]:
children_sample = children_short.sample(1000)
children_sample.head()

Text  Rating  \
226729  A true bookworm! An excellent book to invite c...       4   
657335  Such a lovely picture book! My nephew loves it...       4   
560231  This review was written for The Review Diaries...       4   
315492        This was my favorite book when I was a kid!       3   
422833                      2 copies library, free online       0   

                                                 Title    Author  \
226729                                 Diary of a Worm  children   
657335  Adventures with Andy & Susie: A Stormy Weekend  children   
560231                    Drina Ballerina (Drina, #11)  children   
315492                         Where the Sidewalk Ends  children   
422833                     Around the World in 80 Days  children   

                                                   Genres Language Pages  \
226729  {'children': 877, 'fiction': 79, 'fantasy, par...      eng    40   
657335        {'poetry': 3, 'children': 12, 'fiction': 3}      eng    36   
560231  {'children': 19, 'fiction': 3, 'fantasy, paran...      eng         
315492  {'poetry': 9400, 'children': 10952, 'fiction':...      eng   176   
422833  {'fiction': 61, 'children': 29, 'history, hist...      eng   239   

       Pub_Year  Avg_Rating  Ratings_Count                           User_ID  
226729     2003        4.20          23804  8fd1da8b299d39d0de7b0bffdadbc42e  
657335                 4.26             39  439294fff66f652c84b3842054b2da01  
560231     1995        4.15            133  04c9e0602d1488f9b32da284aaf42c3b  
315492     2002        4.29        1029527  45658f5163313e5236a1f60f2e4e5329  
422833     1989        3.96           4748  03f3dcc940b31eb08df82b39e672c32e

In [66]:
children_sample.shape

(1000, 11)

In [67]:
children_sample.User_ID.value_counts().value_counts()

1    839
2     40
3     13
5      4
4      4
6      1
Name: User_ID, dtype: int64

In [68]:
#To-Do: Fill in null values (page numbers, year, etc.)